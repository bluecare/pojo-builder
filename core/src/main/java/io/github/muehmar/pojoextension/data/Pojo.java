package io.github.muehmar.pojoextension.data;

import ch.bluecare.commons.data.PList;
import java.util.Objects;

public class Pojo {
  private final Name extensionName;
  private final Name pojoName;
  private final PackageName pkg;
  private final PList<PojoField> fields;

  public Pojo(Name extensionName, Name pojoName, PackageName pkg, PList<PojoField> fields) {
    this.extensionName = extensionName;
    this.pojoName = pojoName;
    this.pkg = pkg;
    this.fields = fields;
  }

  public Name getExtensionName() {
    return extensionName;
  }

  public Name getPojoName() {
    return pojoName;
  }

  public PackageName getPackage() {
    return pkg;
  }

  public PList<PojoField> getFields() {
    return fields;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Pojo pojo = (Pojo) o;
    return Objects.equals(extensionName, pojo.extensionName)
        && Objects.equals(pojoName, pojo.pojoName)
        && Objects.equals(pkg, pojo.pkg)
        && Objects.equals(fields, pojo.fields);
  }

  @Override
  public int hashCode() {
    return Objects.hash(extensionName, pojoName, pkg, fields);
  }

  @Override
  public String toString() {
    return "Pojo{"
        + "extensionName="
        + extensionName
        + ", pojoName="
        + pojoName
        + ", pkg="
        + pkg
        + ", fields="
        + fields
        + '}';
  }
}
