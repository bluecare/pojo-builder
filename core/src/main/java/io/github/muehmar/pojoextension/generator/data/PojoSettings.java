package io.github.muehmar.pojoextension.generator.data;

import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.util.Optional;

@PojoExtension
@SuppressWarnings("java:S2160")
public class PojoSettings extends PojoSettingsExtension {
  private final boolean disableSafeBuilder;
  private final Optional<Name> extensionName;

  public PojoSettings(boolean disableSafeBuilder, Optional<Name> extensionName) {
    this.disableSafeBuilder = disableSafeBuilder;
    this.extensionName = extensionName;
  }

  public static PojoSettings defaultSettings() {
    return new PojoSettings(false, Optional.empty());
  }

  public boolean isDisableSafeBuilder() {
    return disableSafeBuilder;
  }

  public boolean isEnableSafeBuilder() {
    return !isDisableSafeBuilder();
  }

  public Name qualifiedExtensionName(Pojo pojo) {
    return pojo.getPkg().qualifiedName(extensionName(pojo));
  }

  public Optional<Name> getExtensionName() {
    return extensionName;
  }

  public Name extensionName(Pojo pojo) {
    return extensionName.orElseGet(
        () -> pojo.getName().map(n -> n.replace(".", "")).append("Extension"));
  }
}
