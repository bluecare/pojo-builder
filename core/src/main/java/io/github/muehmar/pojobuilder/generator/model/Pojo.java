package io.github.muehmar.pojobuilder.generator.model;

import static io.github.muehmar.pojobuilder.Booleans.not;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojobuilder.Strings;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import io.github.muehmar.pojobuilder.exception.PojoBuilderException;
import io.github.muehmar.pojobuilder.generator.model.type.Type;
import java.util.Optional;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class Pojo {
  private static final PList<Name> LETTERS_AZ =
      PList.range(65, 91).map(n -> Character.toString((char) n.intValue())).map(Name::fromString);

  PojoName pojoName;
  PackageName pkg;
  PList<PojoField> fields;
  PList<Constructor> constructors;
  PList<Generic> generics;
  PList<FieldBuilder> fieldBuilders;
  Optional<BuildMethod> buildMethod;

  public PackageName getPackage() {
    return pkg;
  }

  public PList<PojoAndField> getPojoAndFields() {
    return fields.map(f -> new PojoAndField(this, f));
  }

  public PList<Name> getGenericImports() {
    return generics.flatMap(Generic::getUpperBounds).flatMap(Type::getImports);
  }

  public String getDiamond() {
    return generics.nonEmpty() ? "<>" : "";
  }

  public PList<String> getGenericTypeDeclarations() {
    return generics.map(Generic::getTypeDeclaration).map(Name::asString);
  }

  public String getGenericTypeDeclarationSection() {
    return Strings.surroundIfNotEmpty("<", getGenericTypeDeclarations().mkString(", "), ">");
  }

  public String getTypeVariablesSection() {
    return Strings.surroundIfNotEmpty(
        "<", generics.map(Generic::getTypeVariable).mkString(", "), ">");
  }

  public Name getNameWithTypeVariables() {
    return pojoName.getName().append(getTypeVariablesSection());
  }

  public String getTypeVariablesWildcardSection() {
    return Strings.surroundIfNotEmpty("<", generics.map(ignore -> "?").mkString(", "), ">");
  }

  public Optional<MatchingConstructor> findMatchingConstructor() {
    return constructors
        .flatMapOptional(c -> c.matchFields(fields).map(f -> new MatchingConstructor(c, f)))
        .headOption();
  }

  public MatchingConstructor getMatchingConstructorOrThrow() {
    return findMatchingConstructor()
        .orElseThrow(() -> new PojoBuilderException(noMatchingConstructorMessage()));
  }

  private String noMatchingConstructorMessage() {
    return String.format(
        "No matching constructor found for class/record %s."
            + " A constructor should have all the fields as arguments in the order of declaration and matching type,"
            + " where the actual type of a non-required field can be wrapped into an java.util.Optional. Furthermore"
            + "it should be accessible from within the same package, i.e. at least package-private. If a field is"
            + "instantiated in the constructor and not part of the arguments, you can annotate it with @Ignore.",
        getPojoName());
  }

  public Name findUnusedTypeVariableName() {
    final PList<Name> typeVariableNames = generics.map(Generic::getTypeVariable);
    return LETTERS_AZ
        .filter(n -> not(typeVariableNames.exists(n::equals)))
        .headOption()
        .orElseThrow(
            () ->
                new IllegalStateException(
                    "All single-letter type variables already used for generic class "
                        + getPojoName()
                        + "! If this is really a use case and should be supported, please contact the maintainer."));
  }

  public Name findUnusedTypeVariableName(Name preferred) {
    final PList<Name> typeVariableNames = generics.map(Generic::getTypeVariable);
    return LETTERS_AZ
        .cons(preferred)
        .filter(n -> not(typeVariableNames.exists(n::equals)))
        .headOption()
        .orElseThrow(
            () ->
                new IllegalStateException(
                    "All single-letter type variables already used for generic class "
                        + getPojoName()
                        + "! If this is really a use case and should be supported, please contact the maintainer."));
  }
}
