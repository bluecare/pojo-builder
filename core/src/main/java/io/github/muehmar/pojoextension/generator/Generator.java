package io.github.muehmar.pojoextension.generator;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.generator.writer.Writer;
import java.util.function.BiPredicate;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;

public interface Generator<A, B> {
  Writer generate(A data, B settings, Writer writer);

  static <A, B> Generator<A, B> ofWriterFunction(UnaryOperator<Writer> f) {
    return (data, settings, writer) -> f.apply(writer);
  }

  static <A, B> Generator<A, B> emptyGen() {
    return (data, settings, writer) -> writer;
  }

  default Generator<A, B> appendNewLine() {
    return append((UnaryOperator<Writer>) Writer::println);
  }

  default Generator<A, B> append(Generator<A, B> next) {
    final Generator<A, B> self = this;
    return (data, settings, writer) ->
        next.generate(data, settings, self.generate(data, settings, writer));
  }

  default Generator<A, B> append(Generator<A, B> next, int tabs) {
    final Generator<A, B> self = this;
    return (data, settings, writer) ->
        self.generate(data, settings, writer)
            .append(tabs, next.generate(data, settings, writer.empty()));
  }

  default Generator<A, B> append(UnaryOperator<Writer> next) {
    return append((data, settings, writer) -> next.apply(writer));
  }

  default <C> Generator<A, B> append(Generator<C, B> gen, Function<A, C> f) {
    final Generator<A, B> self = this;
    return (data, settings, writer) ->
        gen.generate(f.apply(data), settings, self.generate(data, settings, writer));
  }

  default <C> Generator<A, B> appendList(Generator<C, B> gen, Function<A, PList<C>> f) {
    final Generator<A, B> self = this;
    return (data, settings, writer) ->
        f.apply(data)
            .foldLeft(self, (g, e) -> g.append(gen, ignore -> e))
            .generate(data, settings, writer);
  }

  default Generator<A, B> appendConditionally(BiPredicate<A, B> predicate, Generator<A, B> append) {
    final Generator<A, B> self = this;
    return (data, settings, writer) -> {
      if (predicate.negate().test(data, settings)) {
        return self.generate(data, settings, writer);
      }
      return append(append).generate(data, settings, writer);
    };
  }

  default Generator<A, B> appendConditionally(Predicate<A> predicate, Generator<A, B> append) {
    return appendConditionally((data, settings) -> predicate.test(data), append);
  }
}
