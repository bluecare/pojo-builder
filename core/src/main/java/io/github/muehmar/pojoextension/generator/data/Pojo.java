package io.github.muehmar.pojoextension.generator.data;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import io.github.muehmar.pojoextension.generator.data.PojoExtension.Builder0;
import java.util.Objects;
import java.util.Optional;

@PojoExtension
public class Pojo {
  private final Name name;
  private final PackageName pkg;
  private final PList<PojoField> fields;
  private final PList<Constructor> constructors;

  public Pojo(
      Name name, PackageName pkg, PList<PojoField> fields, PList<Constructor> constructors) {
    this.name = name;
    this.pkg = pkg;
    this.fields = fields;
    this.constructors = constructors;
  }

  public Name getName() {
    return name;
  }

  public Name qualifiedName() {
    return pkg.qualifiedName(name);
  }

  public PackageName getPackage() {
    return pkg;
  }

  public PList<PojoField> getFields() {
    return fields;
  }

  public PList<Constructor> getConstructors() {
    return constructors;
  }

  public Optional<MatchingConstructor> findMatchingConstructor() {
    return constructors
        .flatMapOptional(c -> c.matchFields(fields).map(f -> new MatchingConstructor(c, f)))
        .headOption();
  }

  public Pojo withFields(PList<PojoField> fields) {
    return new Pojo(name, pkg, fields, constructors);
  }

  public Pojo withConstructors(PList<Constructor> constructors) {
    return new Pojo(name, pkg, fields, constructors);
  }

  public static Builder0 newBuilder() {
    return io.github.muehmar.pojoextension.generator.data.PojoExtension.newBuilder();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Pojo pojo = (Pojo) o;
    return Objects.equals(name, pojo.name)
        && Objects.equals(pkg, pojo.pkg)
        && Objects.equals(fields, pojo.fields)
        && Objects.equals(constructors, pojo.constructors);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name, pkg, fields, constructors);
  }

  @Override
  public String toString() {
    return "Pojo{"
        + "name="
        + name
        + ", pkg="
        + pkg
        + ", fields="
        + fields
        + ", constructors="
        + constructors
        + '}';
  }
}
