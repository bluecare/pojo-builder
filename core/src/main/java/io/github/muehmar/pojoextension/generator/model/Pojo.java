package io.github.muehmar.pojoextension.generator.model;

import static io.github.muehmar.pojoextension.Booleans.not;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.Strings;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import io.github.muehmar.pojoextension.exception.PojoBuilderException;
import io.github.muehmar.pojoextension.generator.model.type.Type;
import java.util.Optional;
import lombok.Value;

@Value
@PojoExtension
public class Pojo implements io.github.muehmar.pojoextension.generator.model.PojoExtension {
  private static final PList<Name> LETTERS_AZ =
      PList.range(65, 91).map(n -> Character.toString((char) n.intValue())).map(Name::fromString);

  Name name;
  PackageName pkg;
  PList<PojoField> fields;
  PList<Constructor> constructors;
  PList<Generic> generics;
  PList<FieldBuilder> fieldBuilders;
  Optional<BuildMethod> buildMethod;

  @io.github.muehmar.pojoextension.annotations.Getter("pkg")
  public PackageName getPackage() {
    return pkg;
  }

  public PList<PojoAndField> getPojoAndFields() {
    return fields.map(f -> new PojoAndField(this, f));
  }

  public PList<Name> getGenericImports() {
    return generics.flatMap(Generic::getUpperBounds).flatMap(Type::getImports);
  }

  public String getDiamond() {
    return generics.nonEmpty() ? "<>" : "";
  }

  public PList<String> getGenericTypeDeclarations() {
    return generics.map(Generic::getTypeDeclaration).map(Name::asString);
  }

  public String getGenericTypeDeclarationSection() {
    return Strings.surroundIfNotEmpty("<", getGenericTypeDeclarations().mkString(", "), ">");
  }

  public String getTypeVariablesSection() {
    return Strings.surroundIfNotEmpty(
        "<", generics.map(Generic::getTypeVariable).mkString(", "), ">");
  }

  public Name getNameWithTypeVariables() {
    return name.append(getTypeVariablesSection());
  }

  public String getTypeVariablesWildcardSection() {
    return Strings.surroundIfNotEmpty("<", generics.map(ignore -> "?").mkString(", "), ">");
  }

  public Optional<MatchingConstructor> findMatchingConstructor() {
    return constructors
        .flatMapOptional(c -> c.matchFields(fields).map(f -> new MatchingConstructor(c, f)))
        .headOption();
  }

  public MatchingConstructor getMatchingConstructorOrThrow() {
    return findMatchingConstructor()
        .orElseThrow(() -> new PojoBuilderException(noMatchingConstructorMessage()));
  }

  private String noMatchingConstructorMessage() {
    return String.format(
        "No matching constructor found for class/record %s."
            + " A constructor should have all the fields as arguments in the order of declaration and matching type,"
            + " where the actual type of a non-required field can be wrapped into an java.util.Optional. Furthermore"
            + "it should be accessible from within the same package, i.e. at least package-private. If a field is"
            + "instantiated in the constructor and not part of the arguments, you can annotate it with @Ignore.",
        getName());
  }

  public Name findUnusedTypeVariableName() {
    final PList<Name> typeVariableNames = generics.map(Generic::getTypeVariable);
    return LETTERS_AZ
        .filter(n -> not(typeVariableNames.exists(n::equals)))
        .headOption()
        .orElseThrow(
            () ->
                new IllegalStateException(
                    "All single-letter type variables already used for generic class "
                        + getName()
                        + "! If this is really a use case and should be supported, please contact the maintainer."));
  }

  public Name findUnusedTypeVariableName(Name preferred) {
    final PList<Name> typeVariableNames = generics.map(Generic::getTypeVariable);
    return LETTERS_AZ
        .cons(preferred)
        .filter(n -> not(typeVariableNames.exists(n::equals)))
        .headOption()
        .orElseThrow(
            () ->
                new IllegalStateException(
                    "All single-letter type variables already used for generic class "
                        + getName()
                        + "! If this is really a use case and should be supported, please contact the maintainer."));
  }
}
