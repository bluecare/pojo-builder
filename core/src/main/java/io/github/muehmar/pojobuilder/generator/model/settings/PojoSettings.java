package io.github.muehmar.pojobuilder.generator.model.settings;

import static java.util.Optional.empty;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojobuilder.annotations.FullBuilderFieldOrder;
import io.github.muehmar.pojobuilder.annotations.OptionalDetection;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import io.github.muehmar.pojobuilder.generator.model.ClassAccessLevelModifier;
import io.github.muehmar.pojobuilder.generator.model.Name;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import java.util.Optional;
import lombok.Value;
import lombok.With;

@Value
@With
@PojoBuilder
public class PojoSettings {
  private static final Name CLASS_NAME_PLACEHOLDER = Name.fromString("{CLASSNAME}");
  public static final Name BUILDER_CLASS_POSTFIX = Name.fromString("Builder");
  PList<OptionalDetection> optionalDetections;
  Optional<Name> builderName;
  Optional<Name> builderSetMethodPrefix;
  ClassAccessLevelModifier builderAccessLevel;
  boolean standardBuilderEnabled;
  boolean fullBuilderEnabled;
  FullBuilderFieldOrder fullBuilderFieldOrder;

  public static PojoSettings defaultSettings() {
    return PojoSettingsBuilder.create()
        .optionalDetections(
            PList.of(OptionalDetection.OPTIONAL_CLASS, OptionalDetection.NULLABLE_ANNOTATION))
        .builderAccessLevel(ClassAccessLevelModifier.PUBLIC)
        .standardBuilderEnabled(true)
        .fullBuilderEnabled(true)
        .fullBuilderFieldOrder(FullBuilderFieldOrder.REQUIRED_FIELDS_FIRST)
        .andAllOptionals()
        .builderName(Optional.of(CLASS_NAME_PLACEHOLDER.append(BUILDER_CLASS_POSTFIX)))
        .builderSetMethodPrefix(empty())
        .build();
  }

  public Name qualifiedBuilderName(Pojo pojo) {
    return pojo.getPackage().qualifiedName(builderName(pojo));
  }

  public Name builderName(Pojo pojo) {
    return getNameOrAppend(builderName, BUILDER_CLASS_POSTFIX, pojo);
  }

  private Name getNameOrAppend(Optional<Name> name, Name postfix, Pojo pojo) {
    return name.map(n -> n.replace(CLASS_NAME_PLACEHOLDER, getClassName(pojo)))
        .orElseGet(() -> getClassName(pojo).append(postfix));
  }

  private Name getClassName(Pojo pojo) {
    return pojo.getName().map(n -> n.replace(".", ""));
  }

  public PojoSettings withBuilderNameOpt(Name builderName) {
    return this.withBuilderName(Optional.of(builderName));
  }

  public PojoSettings withBuilderSetMethodPrefixOpt(Name builderSetMethodPrefix) {
    return this.withBuilderSetMethodPrefix(Optional.of(builderSetMethodPrefix));
  }

  public PojoSettings overrideOptionalDetection(
      Optional<PList<OptionalDetection>> optionalDetections) {
    return optionalDetections.map(this::withOptionalDetections).orElse(this);
  }

  public PojoSettings overrideBuilderName(Optional<Name> builderName) {
    return builderName.map(this::withBuilderNameOpt).orElse(this);
  }

  public PojoSettings overrideBuilderSetMethodPrefix(Optional<Name> builderSetMethodPrefix) {
    return builderSetMethodPrefix
        .map(ignore -> this.withBuilderSetMethodPrefix(builderSetMethodPrefix))
        .orElse(this);
  }

  public PojoSettings overrideBuilderAccessLevel(
      Optional<ClassAccessLevelModifier> classAccessLevelModifier) {
    return classAccessLevelModifier.map(this::withBuilderAccessLevel).orElse(this);
  }
}
