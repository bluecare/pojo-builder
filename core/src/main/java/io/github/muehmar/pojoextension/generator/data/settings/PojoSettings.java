package io.github.muehmar.pojoextension.generator.data.settings;

import static io.github.muehmar.pojoextension.generator.data.settings.Ability.ENABLED;
import static io.github.muehmar.pojoextension.generator.data.settings.ExtensionUsage.INHERITED;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.OptionalDetection;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import io.github.muehmar.pojoextension.generator.data.Name;
import io.github.muehmar.pojoextension.generator.data.Pojo;
import java.util.Optional;

@PojoExtension
@SuppressWarnings("java:S2160")
public class PojoSettings extends PojoSettingsBase {
  private static final Name CLASS_NAME_PLACEHOLDER = Name.fromString("{CLASSNAME}");
  public static final Name BUILDER_CLASS_POSTFIX = Name.fromString("Builder");
  public static final Name EXTENSION_IFC_POSTFIX = Name.fromString("Extension");
  public static final Name BASE_CLASS_POSTFIX = Name.fromString("Base");
  private final PList<OptionalDetection> optionalDetections;
  private final ExtensionUsage extensionUsage;
  private final Optional<Name> extensionName;
  private final Optional<Name> builderName;
  private final Optional<Name> baseClassName;
  private final Ability safeBuilderAbility;
  private final Ability equalsHashCodeAbility;
  private final Ability toStringAbility;
  private final Ability withersAbility;
  private final Ability mappersAbility;
  private final Ability baseClassAbility;

  PojoSettings(
      PList<OptionalDetection> optionalDetections,
      ExtensionUsage extensionUsage,
      Optional<Name> extensionName,
      Optional<Name> builderName,
      Optional<Name> baseClassName,
      Ability safeBuilderAbility,
      Ability equalsHashCodeAbility,
      Ability toStringAbility,
      Ability withAbility,
      Ability mapAbility,
      Ability baseClassAbility) {
    this.optionalDetections = optionalDetections;
    this.extensionUsage = extensionUsage;
    this.extensionName = extensionName;
    this.builderName = builderName;
    this.baseClassName = baseClassName;
    this.safeBuilderAbility = safeBuilderAbility;
    this.equalsHashCodeAbility = equalsHashCodeAbility;
    this.toStringAbility = toStringAbility;
    this.withersAbility = withAbility;
    this.mappersAbility = mapAbility;
    this.baseClassAbility = baseClassAbility;
  }

  public static PojoSettings defaultSettings() {
    return PojoSettingsBuilder.create()
        .setOptionalDetections(
            PList.of(OptionalDetection.OPTIONAL_CLASS, OptionalDetection.NULLABLE_ANNOTATION))
        .setExtensionUsage(INHERITED)
        .setSafeBuilderAbility(ENABLED)
        .setEqualsHashCodeAbility(ENABLED)
        .setToStringAbility(ENABLED)
        .setWithersAbility(ENABLED)
        .setMappersAbility(ENABLED)
        .setBaseClassAbility(ENABLED)
        .andAllOptionals()
        .setExtensionName(Optional.of(CLASS_NAME_PLACEHOLDER.append(EXTENSION_IFC_POSTFIX)))
        .setBuilderName(Optional.of(CLASS_NAME_PLACEHOLDER.append(BUILDER_CLASS_POSTFIX)))
        .setBaseClassName(Optional.of(CLASS_NAME_PLACEHOLDER.append(BASE_CLASS_POSTFIX)))
        .build();
  }

  public PList<OptionalDetection> getOptionalDetections() {
    return optionalDetections;
  }

  public ExtensionUsage getExtensionUsage() {
    return extensionUsage;
  }

  public Optional<Name> getExtensionName() {
    return extensionName;
  }

  public Optional<Name> getBuilderName() {
    return builderName;
  }

  public Optional<Name> getBaseClassName() {
    return baseClassName;
  }

  public Ability getSafeBuilderAbility() {
    return safeBuilderAbility;
  }

  public Ability getEqualsHashCodeAbility() {
    return equalsHashCodeAbility;
  }

  public Ability getToStringAbility() {
    return toStringAbility;
  }

  public Ability getWithersAbility() {
    return withersAbility;
  }

  public Ability getMappersAbility() {
    return mappersAbility;
  }

  public Ability getBaseClassAbility() {
    return baseClassAbility;
  }

  public Name qualifiedExtensionName(Pojo pojo) {
    return pojo.getPackage().qualifiedName(extensionName(pojo));
  }

  public Name extensionName(Pojo pojo) {
    return getNameOrAppend(extensionName, EXTENSION_IFC_POSTFIX, pojo);
  }

  public Name qualifiedBuilderName(Pojo pojo) {
    return pojo.getPackage().qualifiedName(builderName(pojo));
  }

  public Name builderName(Pojo pojo) {
    return getNameOrAppend(builderName, BUILDER_CLASS_POSTFIX, pojo);
  }

  public Name qualifiedBaseClassName(Pojo pojo) {
    return pojo.getPackage().qualifiedName(baseClassName(pojo));
  }

  public Name baseClassName(Pojo pojo) {
    return getNameOrAppend(baseClassName, BASE_CLASS_POSTFIX, pojo);
  }

  private Name getNameOrAppend(Optional<Name> name, Name postfix, Pojo pojo) {
    return name.map(n -> n.replace(CLASS_NAME_PLACEHOLDER, getClassName(pojo)))
        .orElseGet(() -> getClassName(pojo).append(postfix));
  }

  private Name getClassName(Pojo pojo) {
    return pojo.getName().map(n -> n.replace(".", ""));
  }

  public boolean createDiscreteBuilder() {
    return safeBuilderAbility.isEnabled();
  }

  public boolean createExtension() {
    return equalsHashCodeAbility.isEnabled()
        || toStringAbility.isEnabled()
        || withersAbility.isEnabled()
        || mappersAbility.isEnabled();
  }

  public boolean createBaseClass() {
    return (equalsHashCodeAbility.isEnabled() || toStringAbility.isEnabled())
        && baseClassAbility.isEnabled();
  }
}
