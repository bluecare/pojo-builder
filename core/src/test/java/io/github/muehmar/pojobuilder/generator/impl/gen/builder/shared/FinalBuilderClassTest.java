package io.github.muehmar.pojobuilder.generator.impl.gen.builder.shared;

import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static io.github.muehmar.pojobuilder.generator.impl.gen.builder.shared.FinalBuilderClass.finalBuilderClass;
import static io.github.muehmar.pojobuilder.generator.impl.gen.builder.standard.StandardBuilderGenerator.CLASS_NAME_FOR_OPTIONAL;
import static io.github.muehmar.pojobuilder.snapshottesting.SnapshotUtil.writerSnapshot;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.generator.Pojos;
import io.github.muehmar.pojobuilder.generator.model.Pojo;
import io.github.muehmar.pojobuilder.generator.model.settings.PojoSettings;
import io.github.muehmar.pojobuilder.snapshottesting.SnapshotTest;
import org.junit.jupiter.api.Test;

@SnapshotTest
class FinalBuilderClassTest {
  private Expect expect;

  @Test
  @SnapshotName("samplePojo")
  void finalBuilderClass_when_samplePojo_then_correctClassOutput() {
    final Generator<Pojo, PojoSettings> generator =
        finalBuilderClass(CLASS_NAME_FOR_OPTIONAL, ignore -> 1);
    final Writer writer =
        generator.generate(Pojos.sample(), PojoSettings.defaultSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("genericPojo")
  void finalOptionalBuilder_when_genericPojo_then_correctClassOutput() {
    final Generator<Pojo, PojoSettings> generator =
        finalBuilderClass(CLASS_NAME_FOR_OPTIONAL, ignore -> 1);
    final Writer writer =
        generator.generate(Pojos.genericSample(), PojoSettings.defaultSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }

  @Test
  @SnapshotName("factoryMethodSample")
  void finalOptionalBuilder_when_factoryMethodSample_then_correctClassOutput() {
    final Generator<Pojo, PojoSettings> generator =
        finalBuilderClass(CLASS_NAME_FOR_OPTIONAL, ignore -> 1);
    final Writer writer =
        generator.generate(
            Pojos.factoryMethodSample(), PojoSettings.defaultSettings(), javaWriter());

    expect.toMatchSnapshot(writerSnapshot(writer));
  }
}
