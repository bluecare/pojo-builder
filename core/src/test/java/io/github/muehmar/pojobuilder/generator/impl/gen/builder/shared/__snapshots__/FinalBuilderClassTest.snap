factoryMethodSample=[
io.github.muehmar.exception.SomeException

public static final class OptBuilder1 {
  private final Builder builder;

  private OptBuilder1(Builder builder) {
    this.builder = builder;
  }

  public io.github.muehmar.Customer build() throws SomeException {
    return builder.build();
  }
}
]


genericPojo=[
java.io.IOException
java.lang.String
java.util.List

public static final class OptBuilder1<T extends List<String>, S> {
  private final Builder<T, S> builder;

  private OptBuilder1(Builder<T, S> builder) {
    this.builder = builder;
  }

  public Customer<T, S> build() throws IOException {
    return builder.build();
  }
}
]


samplePojo=[
java.io.IOException

public static final class OptBuilder1 {
  private final Builder builder;

  private OptBuilder1(Builder builder) {
    this.builder = builder;
  }

  public Customer build() throws IOException {
    return builder.build();
  }
}
]